Support loading assemblies and symbols into debugger via callback

captainsafia opened the pull request:
This PR adds support for loading assemblies/symbols into the debugging via an `assembly_loaded` callback as opposed to the current model which loads assemblies/symbols when the application is initialized via the `loaded_files` singleton.

Changes in this PR include:
- Splitting up debugger-test and lazy-debugger-test projects to support testing
- Adding `mono_wasm_add_files` method to signal debugger from client using console.log
- Added assembly_load callback to WASM debugger agent
- Add handling for `ADD_ASSEMBLY_PDB` event in MonoProxy

Validated via unit tests and StandaloneApp in Blazor. LazyAssemblyLoader needs to be modified to support this feature. I'll submit a PR to the aspnetcore repo for that.

Commit:
Add test setup for debugging lazy-loaded assemblies

Commit:
Add mono_wasm_add_lazy_load_files callback

Commit:
Rely on assembly_load callback to register PDBs

lambdageek commented on src/mono/mono/mini/mini-wasm-debugger.c:
```diff
@@ -64,11 +65,13 @@ extern void mono_wasm_add_properties_var (const char*, gint32);
 extern void mono_wasm_add_array_item (int);
 extern void mono_wasm_set_is_async_method (guint64);
 extern void mono_wasm_add_typed_value (const char *type, const char *str_value, double value);
+extern void mono_wasm_add_files(const char *assembly_data, guint32 assembly_len, const char *pdb_data, guint32 pdb_len);
```
nit: space before '('


```suggestion
extern void mono_wasm_add_files (const char *assembly_data, guint32 assembly_len, const char *pdb_data, guint32 pdb_len);
```

captainsafia replied:
Thanks! I've resolved your comments because I've addressed the issues on my local branch. Will push once I've gotten more feedback from others.

lambdageek commented on src/mono/mono/mini/mini-wasm-debugger.c:
```diff
@@ -64,11 +65,13 @@ extern void mono_wasm_add_properties_var (const char*, gint32);
 
 static void describe_object_properties_for_klass (void *obj, MonoClass *klass, gboolean isAsyncLocalThis, int gpflags);
 static void handle_exception (MonoException *exc, MonoContext *throw_ctx, MonoContext *catch_ctx, StackFrameInfo *catch_frame);
+static void assembly_load(MonoProfiler *prof, MonoAssembly *assembly);
```
nit
```suggestion
static void assembly_load (MonoProfiler *prof, MonoAssembly *assembly);
```

lambdageek commented on src/mono/mono/mini/mini-wasm-debugger.c:
```diff
@@ -360,6 +363,7 @@ mono_wasm_debugger_init (void)
 	mono_profiler_set_jit_done_callback (prof, jit_done);
 	//FIXME support multiple appdomains
 	mono_profiler_set_domain_loaded_callback (prof, appdomain_load);
+	mono_profiler_set_assembly_loaded_callback(prof, assembly_load);
```
nit:
* mono uses tabs, not spaces, with tab width = 8
* space before '('
```suggestion
	mono_profiler_set_assembly_loaded_callback (prof, assembly_load);
```

lambdageek commented on src/mono/mono/mini/mini-wasm-debugger.c:
```diff
@@ -456,6 +460,23 @@ get_object_id(MonoObject *obj)
 }
 
+static void
+assembly_load(MonoProfiler *prof, MonoAssembly *assembly)
```
nit
```suggestion
assembly_load (MonoProfiler *prof, MonoAssembly *assembly)
```

lambdageek commented on src/mono/mono/mini/mini-wasm-debugger.c:
```diff
@@ -456,6 +460,23 @@ get_object_id(MonoObject *obj)
+static void
+assembly_load(MonoProfiler *prof, MonoAssembly *assembly)
+{
+    DEBUG_PRINTF(1, "Running assembly_loaded callback for %s", assembly->aname.name);
+    MonoImage *assembly_image = assembly->image;
+    MonoImage *pdb_image = NULL;
+    MonoDebugHandle *handle = mono_debug_get_handle(assembly_image);
+    MonoPPDBFile *ppdb = handle->ppdb;
+    if (ppdb)
+    {
+        pdb_image = mono_ppdb_get_image(ppdb);
+        mono_wasm_add_files(assembly_image->raw_data, assembly_image->raw_data_len, pdb_image->raw_data, pdb_image->raw_data_len);
+        return;
+    }
+    mono_wasm_add_files(assembly_image->raw_data, assembly_image->raw_data_len, NULL, 0);
```
nit:
* brace for `if` goes on the same line as the `if`
* spaces vs tabs
* space before '('
```suggestion
	DEBUG_PRINTF(1, "Running assembly_loaded callback for %s", assembly->aname.name);
	MonoImage *assembly_image = assembly->image;
	MonoImage *pdb_image = NULL;
	MonoDebugHandle *handle = mono_debug_get_handle (assembly_image);
	MonoPPDBFile *ppdb = handle->ppdb;
	if (ppdb) {
		pdb_image = mono_ppdb_get_image (ppdb);
		mono_wasm_add_files (assembly_image->raw_data, assembly_image->raw_data_len, pdb_image->raw_data, pdb_image->raw_data_len);
		return;
	}
	mono_wasm_add_files (assembly_image->raw_data, assembly_image->raw_data_len, NULL, 0);
```

lambdageek reviewed (commented):
For the .c files, please follow the [Mono coding guidelines](https://www.mono-project.com/community/contributing/coding-guidelines/)

captainsafia committed:
Address feedback on formatting

captainsafia committed:
Fix spacing on method invocations

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:
```diff
@@ -735,6 +735,29 @@ class DebugItem
+            }
+            catch (Exception e)
+            {
+                logger.LogDebug($"Failed to load assembly: ({e.Message})");
```
`yield break;` after this?

captainsafia replied:
Good call! Done.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -821,6 +823,46 @@ async Task<bool> OnJSEventRaised(SessionId sessionId, JObject eventArgs, Cancell
+            }
+            catch (Exception e)
+            {
+                logger.LogDebug($"Failed to load assemblies and PBDs: {e}");
```
`PBDs`->`PDBs`

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -821,6 +823,46 @@ async Task<bool> OnJSEventRaised(SessionId sessionId, JObject eventArgs, Cancell
             }
         }
 
+        async Task<bool> HandleAddAssemblyPdb(SessionId sessionId, JObject eventArgs, CancellationToken token)
```
Maybe `OnAssemblyLoadedJSEvent`?

radical commented on src/mono/mono/mini/mini-wasm-debugger.c:
```diff
@@ -456,6 +460,22 @@ get_object_id(MonoObject *obj)
+static void
+assembly_load (MonoProfiler *prof, MonoAssembly *assembly)
+{
+	DEBUG_PRINTF (1, "Running assembly_loaded callback for %s", assembly->aname.name);
```
```suggestion
	DEBUG_PRINTF (2, "assembly_loaded callback called for %s\n", assembly->aname.name);
```

radical commented on src/mono/mono/mini/mini-wasm-debugger.c:
```diff
@@ -456,6 +460,22 @@ get_object_id(MonoObject *obj)
+	DEBUG_PRINTF (1, "Running assembly_loaded callback for %s", assembly->aname.name);
+	MonoImage *assembly_image = assembly->image;
+	MonoImage *pdb_image = NULL;
+	MonoDebugHandle *handle = mono_debug_get_handle (assembly_image);
```
Log an error, and return `if (!handle)`

captainsafia replied:
Done.

radical commented on src/mono/mono/mini/mini-wasm-debugger.c:
```diff
@@ -456,6 +460,22 @@ get_object_id(MonoObject *obj)
+	MonoImage *pdb_image = NULL;
+	MonoDebugHandle *handle = mono_debug_get_handle (assembly_image);
+	MonoPPDBFile *ppdb = handle->ppdb;
+	if (ppdb) {
+		pdb_image = mono_ppdb_get_image (ppdb);
+		mono_wasm_add_files (assembly_image->raw_data, assembly_image->raw_data_len, pdb_image->raw_data, pdb_image->raw_data_len);
+		return;
+	}
+	mono_wasm_add_files (assembly_image->raw_data, assembly_image->raw_data_len, NULL, 0);
+}
```
```suggestion
	if (ppdb) {
		pdb_image = mono_ppdb_get_image (ppdb);
		mono_wasm_add_files (assembly_image->raw_data, assembly_image->raw_data_len, pdb_image->raw_data, pdb_image->raw_data_len);
	} else {
	    mono_wasm_add_files (assembly_image->raw_data, assembly_image->raw_data_len, NULL, 0);
	}
}
```

radical commented on src/mono/mono/mini/mini-wasm-debugger.c:
```diff
@@ -64,11 +65,13 @@ extern void mono_wasm_add_properties_var (const char*, gint32);
 extern void mono_wasm_add_array_item (int);
 extern void mono_wasm_set_is_async_method (guint64);
 extern void mono_wasm_add_typed_value (const char *type, const char *str_value, double value);
+extern void mono_wasm_add_files (const char *assembly_data, guint32 assembly_len, const char *pdb_data, guint32 pdb_len);
```
maybe `mono_wasm_assembly_loaded_event`

captainsafia replied:
Opted for `mono_wasm_asm_loaded` since longer method names cause issues with the function registration. @thaystg was able to help me resolve this earlier when the method name was `mono_wasm_add_lazy_loaded_files`.

thaystg replied:
No no, the problem wasn't the function name, was that the function was inside MONO variable. You can use the original name if you want

radical commented on src/mono/mono/mini/mini-wasm-debugger.c:
```diff
@@ -64,11 +65,13 @@ extern void mono_wasm_add_properties_var (const char*, gint32);
 
 static void describe_object_properties_for_klass (void *obj, MonoClass *klass, gboolean isAsyncLocalThis, int gpflags);
 static void handle_exception (MonoException *exc, MonoContext *throw_ctx, MonoContext *catch_ctx, StackFrameInfo *catch_frame);
+static void assembly_load (MonoProfiler *prof, MonoAssembly *assembly);
```
`assembly_loaded`

radical commented on src/mono/wasm/runtime/library_mono.js:
```diff
@@ -2252,6 +2252,16 @@ var MonoSupportLib = {
+		const pdb_data = pdb_ptr ? new Uint8Array(Module.HEAPU8.buffer, pdb_ptr, pdb_len) : null;
+		MONO.mono_wasm_raise_debug_event({
+			eventName: 'ADD_ASSEMBLY_PDB',
+			assembly_data: Array.from(assembly_data),
```
This returns an array, which gets stringified, so `[4,12,2, ... ]`. It would be better to return a base64 string instead.
I was trying ` const assembly_b64 = btoa (String.fromCharCode (...assembly_data));`, but that ends up exceeding the call stack size. We could encode that manually in a loop, but maybe you know of some better way to do this?

The good way to do this would be to use cdp's IO api, but that might be too much new code close to rc2. I will try that locally anyway.

radical replied:
You can add this to `$MONO` object - https://gist.github.com/radical/00c86f5617fb8f26db5a3b6b5ff58361 . It's based on JSIL code, thanks to @kg .

And use it like:
```
		const assembly_data = new Uint8Array(Module.HEAPU8.buffer, assembly_ptr, assemnly_len);
		const assembly_b64 = MONO._base64Converter.toBase64StringImpl(assembly_data);

		let pdb_b64;
		if (pdb_ptr) {
			const pdb_data = new Uint8Array(Module.HEAPU8.buffer, pdb_ptr, pdb_len);
			pdb_b64 = MONO._base64Converter.toBase64StringImpl(pdb_data);
		}
```

I tried it out, and it works fine. Once that is on here, we can ask her to review it.

captainsafia replied:
Yeah! I spent a bit of time playing with different approaches here but wasn't able to find something that worked or wasn't unbearably slow. I'll integrate these changes and see how that goes.

captainsafia replied:
The change has been integrated. I see this is a code change pulled out from another codebase. Do we need to add additional tests for it here?

Also: does it make sense to put this under the BINDING namespace?

thaystg commented on src/mono/wasm/runtime/library_mono.js:
```diff
@@ -2252,6 +2252,16 @@ var MonoSupportLib = {
 	},
+
+	mono_wasm_add_files: function (assembly_ptr, assemnly_len, pdb_ptr, pdb_len) {
+		const assembly_data = new Uint8Array(Module.HEAPU8.buffer, assembly_ptr, assemnly_len);
```
Why it's not a good idea to register all the assemblies using this new infrastructure?

radical replied:
This will send the assembly as a base64 string, which would be a bigger "download". But also, I would want to do a lot more testing, and checking that moving assembly loading for the proxy, entirely to this technique, works well, and is a good idea. So, I think we should use minimize it's use for now.

thaystg replied:
Okay, I agree that for RC2, we should minimize the use of the new technique. Thanks, Ankit.

captainsafia replied:
That seems prudent. I assume .NET 6 is the earlier we'd migrate to this then?

> a lot more testing, and checking that moving assembly loading for the proxy, entirely to this technique, works well, and is a good idea.

Curious: what's your intuition on why this would not be a good idea?

thaystg reviewed (approved):
LGTM. I tested locally using this branch and looks like it's working.
I agree with Ankit's suggestion, unless the one that I commented.

thaystg commented on src/mono/wasm/debugger/tests/debugger-test/debugger-test.cs:
```diff
@@ -518,3 +518,15 @@ public static void run()
+        byte[] asm_bytes = Convert.FromBase64String(asm_base64);
+        byte[] pdb_bytes = Convert.FromBase64String(pdb_base64);
+
+        var loadedAssembly = System.Runtime.Loader.AssemblyLoadContext.Default.LoadFromStream(new System.IO.MemoryStream(asm_bytes), new System.IO.MemoryStream(pdb_bytes));
```
It works, I have already tested, and it's a good idea to add it in the tests.

captainsafia replied:
> Is that a valid case for us?

At least for the Blazor scenario, we'll always load into the default context so there's no valid case there.

> It works, I have already tested, and it's a good idea to add it in the tests.

Sure. I can add a test.

captainsafia replied:
So, this works-ish. When loaded into a separate context, the callback is triggered, PDBs are loaded into the DebugStore, and scripts are parsed.

However, setting breakpoints fails because we currently do a lookup for the assembly in the default context ([ref](https://github.com/dotnet/runtime/blob/master/src/mono/mono/mini/mini-wasm-debugger.c#L513)).

Not sure if we want to address this now considering we don't anticipate loading into the non-default context for the Blazor scenario.

I imagine supporting this will require modifying the `mono_wasm_set_breakpoint` method to support looks up in multiple contexts.

lambdageek replied:
@captainsafia (or Ankit or Thays) could you create an issue for MonoVM to support for setting breakpoints in non-default ALCs?

I think it probably affects the non-wasm scenarios too.  Actually I wonder what happens if we load the same assembly into multiple ALCs

thaystg replied:
I can create the issue.

Commit:
Address second round of feedback

Commit:
Transport assembly data as base64 string

Commit:
Revert automated whitespace changes

captainsafia commented on src/mono/wasm/debugger/DebuggerTestSuite/Tests.cs:
```diff
@@ -1850,6 +1851,51 @@ JObject FindFrame(JObject pause_location, string function_name)
                     ?.FirstOrDefault();
 
+        [Fact]
+        public async Task DebugLazyLoadedAssembly()
```
Done.

captainsafia commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -925,38 +967,6 @@ async Task<DebugStore> LoadStore(SessionId sessionId, CancellationToken token)
             if (Interlocked.CompareExchange(ref context.store, new DebugStore(logger), null) != null)
                 return await context.Source.Task;
 
-            try
```
Restored.

radical committed:
[wasm][debugger] Fix some issues with lazy loading assemblies

- One of the problems I found was that we were sending lot of duplicate
assemblies.
    - this is a problem because we are converting them to base64, then
    sending them over the wire, unnecessarily.
    - Also, the proxy wasn't correctly dealing with them, and ended up
    processing them, and then sending out duplicate `scriptParsed`
    events!

- So, we try to solve this at the `mono_wasm_asm_loaded` point:
    - issue: the `MONO.mono_wasm_runtime_is_ready` check wasn't enough
    - What we want to do is to send the events only for assemblies that
      were not in the bundle, list of which we can get in
      `MONO.loaded_files`.
    - but that has the filename
    - But we don't seem to be getting the filename from the MonoImage,
      so we use the assembly name as a workaround.

    - So, as a heuristic (and trying not too add too much new stuff this
    close to rc2!), we check if the assembly name begins with `System.`,
    or `Microsoft.`, and if it does then we look for `$assembly_name.dll`
    in the the loaded filenames.
        - If found, then we can skip these!

- The other point in the Proxy:
    - where we avoid adding duplicate assemblies if we already have one
    keyed by the assembly name.

radical commented:
I was trying out this branch, and found some issues. I have pushed a fix for those, and some tests to - https://github.com/radical/runtime/commit/f7bfca9a0f519addbc31ef2637197bbf176f2bd6 . You can add this to this PR.
Then it needs another round of review.

And I would suggest - make sure to `run-debugger-tests` locally, and just try some basic sanity check by debugging some app.

radical commented:
@thaystg that commit needs a review, but it can wait till when it gets added here. I didn't want to push without asking.

captainsafia commented:
> I was trying out this branch, and found some issues. I have pushed a fix for those, and some tests to - radical@f7bfca9 . You can add this to this PR.

Can you list out what the issues where? I ran `run-debugger-tests` locally and got 3 test failures that seemed to be flakes (e.g. `System.AggregateException : One or more errors occurred. (A task was canceled.)` from the WebSocket listener). Would be good to know what you found.

radical commented:
> > I was trying out this branch, and found some issues. I have pushed a fix for those, and some tests to - radical@f7bfca9 . You can add this to this PR.
> 
> 
> 
> Can you list out what the issues where? I ran `run-debugger-tests` locally and got 3 test failures that seemed to be flakes (e.g. `System.AggregateException : One or more errors occurred. (A task was canceled.)` from the WebSocket listener). Would be good to know what you found.

I mentioned that in the commit message.

radical commented:
Shouldn't be flaky. Make sure to run 'make build' before running the tests. If you are still getting the errors, then could you share the full log for those?

radical commented:
@lewing could you take a look at this? This has some changes in `DebugStore`, and `MonoProxy`.

CoffeeFlux commented on src/mono/mono/mini/mini-wasm-debugger.c:
```diff
@@ -65,7 +65,7 @@ extern void mono_wasm_add_properties_var (const char*, gint32);
 extern void mono_wasm_set_is_async_method (guint64);
 extern void mono_wasm_add_typed_value (const char *type, const char *str_value, double value);
-extern void mono_wasm_asm_loaded (const char *assembly_data, guint32 assembly_len, const char *pdb_data, guint32 pdb_len);
+extern void mono_wasm_asm_loaded (const char* asm_name, const char *assembly_data, guint32 assembly_len, const char *pdb_data, guint32 pdb_len);
```
```suggestion
extern void mono_wasm_asm_loaded (const char *asm_name, const char *assembly_data, guint32 assembly_len, const char *pdb_data, guint32 pdb_len);
```

captainsafia replied:
Thanks! Fixed this.

CoffeeFlux reviewed (commented):
Quick nit

Commit:
Fix spacing in function definition

lewing commented on src/mono/mono/mini/mini-wasm-debugger.c:
```diff
@@ -456,6 +460,29 @@ get_object_id(MonoObject *obj)
+		pdb_image = mono_ppdb_get_image (ppdb);
+	}
+
+	if (pdb_image) {
```
if the assembly has and embedded pdb are we sending it over the wire twice?

lewing replied:
We probably also want to either pass the value of mono_has_pdb_checksum as an arg to mono_wasm_asm_loaded or skip calling it entirely if it is false.

radical replied:
> if the assembly has and embedded pdb are we sending it over the wire twice?

good point. @thaystg can you take these?

thaystg replied:
I'm working on it, pushing a commit soon.

thaystg replied:
Yes, we are sending twice, inside the dll and the data outside the dll, we don't need to send the pdb outside if it's embedded.

thaystg replied:
@lewing Now we check if it's an embedded PPDB and in this case send only the assembly, and we check if it's a assembly without debug information and in this case we don't send anything.

lewing commented on src/mono/wasm/debugger/DebuggerTestSuite/Tests.cs:
```diff
@@ -1850,6 +1851,70 @@ JObject FindFrame(JObject pause_location, string function_name)
+        }
+
+        [Fact]
+        public async Task DebugLazyLoadedAssemblyWithoutPdb()
```
Please add a test with an assembly with embedded ppdb data.

thaystg replied:
Working on it.

thaystg replied:
Done

lewing reviewed (commented):
This is looking good but I still thinking through the initial load logic and the duplication we're seeing there.  More comments in a few minutes.

radical committed:
[wasm][debugger] Handle errors in downloading from symbol servers

radical committed:
[wasm][debugger] MonoProxy: Don't skip assembly if no pdb is available

- this caused such assemblies not getting added `DebugStore.assemblies`
- which broke download symbols on demand, because it couldn't find the
corresponding assembly in the list

radical committed:
[wasm][debugger] Use the correct assembly name for the search

radical reviewed (commented):
Since, I contributed non-trivially to this PR, I'll skip approving it :)
(I can't see how to dismiss my earlier review!)

lewing commented on src/mono/wasm/runtime/library_mono.js:
```diff
@@ -2252,6 +2359,57 @@ var MonoSupportLib = {
+
+		const assembly_name_str = assembly_name !== 0 ? Module.UTF8ToString(assembly_name) : '';
+		if (assembly_name !== 0) {
+			if ((assembly_name_str.startsWith("System.") || assembly_name_str.startsWith("Microsoft."))
```
I don't like this name hack.  driver.c has a linked list of the preloaded assemblies can't we dedup off that by address?

captainsafia replied:
@radical made this change so I'll let him speak on the trade-offs here.

captainsafia replied:
Chatted with @thaystg about this. She suspects it wasn't intentional. I'll make the change here.

captainsafia replied:
I've added a new exported function in driver.c that searches for assemblies already loaded by name and updated the check here to leverage that.

radical replied:
Glad to be rid of the hack! 👍

lewing commented on src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:
```diff
@@ -735,6 +738,36 @@ class DebugItem
 
+        public IEnumerable<SourceFile> Add(SessionId sessionId, byte[] assembly_data, byte[] pdb_data)
+        {
+            var resolver = new DefaultAssemblyResolver();
```
Is there any reason not to make the resolver a readonly member?  I think it might avoid some extra searching overhead.

captainsafia replied:
I don't see any reason not too. I can make this change.

lewing replied:
The lifetime of the DebugStore is a bit different than the runtime assemblies so the reload resolution might be slightly different?

captainsafia replied:
From reading the code, I think it should be fine. Initializing the resolver once will allow both `Add` and `Load` to share the same cache, which I believe is what we want here.

thaystg committed:
Fixing and changes things related to Embedded Pdb.

lewing commented:
Okay, I think this is quite close now. If the debugger attaches after the events have fired it still won't be able to see the lazy loaded assemblies but that isn't the common case and can be tracked in a separate issue.

Commit:
Update check for loaded assemblies and initialize resolver once

lambdageek commented on src/mono/mono/mini/mini-wasm-debugger.c:
```diff
@@ -456,6 +460,30 @@ get_object_id(MonoObject *obj)
+		mono_wasm_asm_loaded (
+			assembly_image->assembly_name, assembly_image->raw_data, assembly_image->raw_data_len,
+			pdb_image->raw_data, pdb_image->raw_data_len);
+	} else if (mono_has_pdb_checksum (assembly_image->raw_data, assembly_image->raw_data_len)) { //if it's a release assembly we don't need to send to DebuggerProxy
```
Is it always the case that `pdb_image` will be NULL for files that have a checksum?  Would it make sense to check for the checksum first?

thaystg replied:
Yes, the case where we are doing an assembly load of some file that we don't have the pdb locally and it's not embedded either, but could have the pdb in a symbol server for example?

thaystg replied:
I will change the order of the checks.

radical replied:
So, just for my understanding, is this what you mean?

```
if (!mono_has_pdb_checksum) return; // nothing to be sent

if (pdb_image) { .. send asm + pdb .. }
else { .. send just asm.. }
```

lambdageek replied:
oh I think I misunderstood the code and what a checksum is.

I thought that there is a checksum present only on "official" released images for which we can expect to find the pdb on a symbol server.

What I was thinking:
```
  if (mono_has_pdb_checksum) { send asm /* don't send pdb, can get it from a symbol server */ }
  else if (pdb_image) { send asm + pdb /* no pdb on symbol server, but we have it in the budle */ }
  else { send asm only; /* no pdb on server or in bundle, just send the assembly */ }
```

thaystg replied:
I rewrote the code, please check if it's clearer?

thaystg replied:
if (!mono_has_pdb_checksum) return; // nothing to be sent

if (pdb_image && !ppdb_embedded) { .. send asm + pdb .. }
else { .. send just asm.. }

thaystg committed:
Changing the order of the checks.

CoffeeFlux commented on src/mono/mono/metadata/debug-mono-ppdb.c:
```diff
@@ -442,7 +446,14 @@ mono_ppdb_lookup_location (MonoDebugMethodInfo *minfo, uint32_t offset)
 mono_ppdb_get_image (MonoPPDBFile *ppdb)
 {
-    return  ppdb->image;
+    return ppdb->image;
```
I think the diff here is an indentation change (tabs->spaces)? The changes here as a whole seems to use spaces :(

thaystg replied:
Fixed.

thaystg committed:
Fix indentation.

radical commented on src/mono/wasm/runtime/library_mono.js:
```diff
@@ -2252,6 +2359,36 @@ var MonoSupportLib = {
+
+		// And for assemblies that have not already been loaded
+		const assembly_name_str = assembly_name !== 0 ? Module.UTF8ToString(assembly_name).concat('.dll') : '';
+		if (this.mono_wasm_assembly_already_added(assembly_name_str))
```
What about: `foo.dll, assembly name "Bar"`? We check for `Bar.dll`. IIUC, this doesn't break anything though, since we'll just end up sending this assembly anyway, and the proxy can dedupe it based on the assembly_name.

I think @lewing was suggesting that may be we can compare the image/data pointer? Would that make sense? Is that accessible in driver.c?

captainsafia replied:
> I think @lewing was suggesting that may be we can compare the image/data pointer? Would that make sense? Is that accessible in driver.c?

Yeah, there is an `entry->assembly.data` property but the addresses don't seem to be the same there so the comparison always fails.

radical committed:
[wasm][debugger][tests] Improve tests a bit

.. which includes setting breakpoint using a url, before loading the
dynamic assembly. This essentially tests that the bp request gets resolved
automatically when the assembly gets loaded.

- And add the embedded pdb test for `DuplicateAssemblyLoadedEvent*`
tests

lewing commented on src/mono/wasm/runtime/driver.c:
```diff
@@ -205,6 +206,22 @@ mono_wasm_add_assembly (const char *name, const unsigned char *data, unsigned in
 }
 
+EMSCRIPTEN_KEEPALIVE int
+mono_wasm_assembly_already_added (const char *assembly_name)
```
the comparison could probably be by raw address but this should be fine

lewing reviewed (approved):
This looks good, I'd like to see a follow up issue for tracking the lazy loaded assemblies for the attach scenario but getting this in will enable debugging of late loaded assemblies which is a pretty important part of the developer experience.

radical commented:
Umm.. just added a few test changes

radical commented:
Just wanted to ensure we test pending bp request resolution, for the lazy loaded assemblies.

lewing commented:
Are we sending these large events even when a debugger isn't attached?

radical commented:
> Are we sending these large events even when a debugger isn't attached?

How do we check that a debugger is connected, on the app side?
If we have to do it in the proxy, then we can send an event without the base64 strings, and let it fetch if it wants to.

lewing commented:
> > Are we sending these large events even when a debugger isn't attached?
> 
> How do we check that a debugger is connected, on the app side?
> If we have to do it in the proxy, then we can send an event without the base64 strings, and let it fetch if it wants to.

There are a couple of ways, but we could not send events until get_loaded_files is called.  That would could also work as a flag for the late attach case.

radical commented:
> > > Are we sending these large events even when a debugger isn't attached?
> > 
> > 
> > How do we check that a debugger is connected, on the app side?
> > If we have to do it in the proxy, then we can send an event without the base64 strings, and let it fetch if it wants to.
> 
> There are a couple of ways, but we could not send events until get_loaded_files is called since it should never be called by anything but the debugger. That would could also work as a flag for the late attach case

though it would continue to send in case of attach-debug-then-detach case?

lewing commented:
> 
> though it would continue to send in case of attach-debug-then-detach case?

I don't think we need to worry a lot about optimizing that case

captainsafia commented:
/backport to release/5.0-rc2

Commit:
Merge remote-tracking branch 'origin/master' into safia/ll-dbg

Commit:
Add back types to DebugStore

Commit:
Comply with new code style requirements
